File: apps/portal/src/lib/api/schemas/projectOverview.ts
import { z } from 'zod';

import { ActivityTimelineBucketSchema } from './activity';
import { AttachmentSchema } from './attachments';
import { CertificateSchema } from './certificates';
import { DeadlineSchema, ReadinessItemSchema } from './deadlines';
import { FindingSchema } from './findings';
import { ProjectMemberSchema, ProjectSchema } from './projects';
import { ReportSchema } from './reports';

// ---------------------------------------------------------------------------
// Project overview — the BFF aggregate behind GET /projects/{id}/overview.
//
// One call assembles the whole project-detail dashboard (header KPIs, the
// completeness donut, capped previews + exact counts per resource, members and
// the activity trend), replacing ~10 separate cold-load requests. Every block
// reuses the existing per-resource Read schemas so the portal shares its types.
// Mirrors apps/api/.../schemas/project_overview.py + deadlines/completeness.py.
// ---------------------------------------------------------------------------

// --- Completeness donut (mirror of deadlines/completeness.py) --------------

export const CompletenessSegmentSchema = z.object({
  category: z.string(),
  filled: z.number(),
  total: z.number(),
});
export type CompletenessSegment = z.infer<typeof CompletenessSegmentSchema>;

export const DossierBlockSchema = z.object({
  filled: z.number(),
  total: z.number(),
  pct: z.number(),
  optional_filled: z.number(),
  optional_total: z.number(),
  segments: z.array(CompletenessSegmentSchema),
  items: z.array(ReadinessItemSchema),
});
export type DossierBlock = z.infer<typeof DossierBlockSchema>;

export const FindingsRingBlockSchema = z.object({
  total: z.number(),
  complete: z.number(),
  by_status: z.record(z.number()),
});
export type FindingsRingBlock = z.infer<typeof FindingsRingBlockSchema>;

export const DeadlinesRingBlockSchema = z.object({
  total: z.number(),
  met: z.number(),
  pending: z.number(),
  overdue: z.number(),
});
export type DeadlinesRingBlock = z.infer<typeof DeadlinesRingBlockSchema>;

export const CompletenessBlockSchema = z.object({
  overall_filled: z.number(),
  overall_total: z.number(),
  overall_pct: z.number(),
  dossier: DossierBlockSchema,
  findings: FindingsRingBlockSchema,
  deadlines: DeadlinesRingBlockSchema,
});
export type CompletenessBlock = z.infer<typeof CompletenessBlockSchema>;

// --- Per-resource preview blocks (count + capped preview) ------------------

export const OverviewFindingsBlockSchema = z.object({
  count: z.number(),
  // open + in_progress (the "still needs work" count).
  open: z.number(),
  preview: z.array(FindingSchema),
});
export type OverviewFindingsBlock = z.infer<typeof OverviewFindingsBlockSchema>;

export const OverviewCertificatesBlockSchema = z.object({
  count: z.number(),
  expired: z.number(),
  expiring_soon: z.number(),
  preview: z.array(CertificateSchema),
});
export type OverviewCertificatesBlock = z.infer<typeof OverviewCertificatesBlockSchema>;

export const OverviewAttachmentsBlockSchema = z.object({
  count: z.number(),
  preview: z.array(AttachmentSchema),
});
export type OverviewAttachmentsBlock = z.infer<typeof OverviewAttachmentsBlockSchema>;

export const OverviewReportsBlockSchema = z.object({
  count: z.number(),
  preview: z.array(ReportSchema),
});
export type OverviewReportsBlock = z.infer<typeof OverviewReportsBlockSchema>;

// `preview` is the full (small) deadline list — seeds the deadlines tab.
export const OverviewDeadlinesBlockSchema = z.object({
  total: z.number(),
  met: z.number(),
  overdue: z.number(),
  preview: z.array(DeadlineSchema),
});
export type OverviewDeadlinesBlock = z.infer<typeof OverviewDeadlinesBlockSchema>;

export const OverviewStatsSchema = z.object({
  deadlines_met: z.number(),
  deadlines_total: z.number(),
  attachments_count: z.number(),
  // Dossier-only required percentage (the header's HOLDBACK chip).
  holdback_pct: z.number(),
  delivery_days_remaining: z.union([z.number(), z.null()]),
});
export type OverviewStats = z.infer<typeof OverviewStatsSchema>;

export const ProjectOverviewSchema = z.object({
  project: ProjectSchema,
  completeness: CompletenessBlockSchema,
  stats: OverviewStatsSchema,
  findings: OverviewFindingsBlockSchema,
  certificates: OverviewCertificatesBlockSchema,
  attachments: OverviewAttachmentsBlockSchema,
  reports: OverviewReportsBlockSchema,
  deadlines: OverviewDeadlinesBlockSchema,
  members: z.array(ProjectMemberSchema),
  activity_timeline: z.array(ActivityTimelineBucketSchema),
});
export type ProjectOverview = z.infer<typeof ProjectOverviewSchema>;

File: apps/api/src/bimdossier_api/schemas/project_overview.py
"""Aggregate response for the project-detail dashboard.

One `GET /projects/{id}/overview` call replaces the ~10 the dashboard used to
fire on a cold load. Every block is built from the existing per-resource Read
schemas (no new row types) so the portal reuses its types; previews are capped
(`OVERVIEW_PREVIEW_LIMIT`) and counts come from cheap aggregates, keeping the
payload bounded regardless of project size.
"""

from pydantic import BaseModel

from bimdossier_api.deadlines.completeness import CompletenessBlock
from bimdossier_api.schemas.activity import ActivityTimelineBucket
from bimdossier_api.schemas.attachment import AttachmentRead
from bimdossier_api.schemas.certificate import CertificateRead
from bimdossier_api.schemas.deadline import DeadlineRead
from bimdossier_api.schemas.finding import FindingRead
from bimdossier_api.schemas.project import ProjectMemberRead, ProjectRead
from bimdossier_api.schemas.report import ReportResponse


class FindingsBlock(BaseModel):
    count: int
    # open + in_progress (the "still needs work" count).
    open: int
    preview: list[FindingRead]


class CertificatesBlock(BaseModel):
    count: int
    # valid_until < today.
    expired: int
    # 0..30 days until valid_until (matches the portal's EXPIRY_WARNING_DAYS).
    expiring_soon: int
    preview: list[CertificateRead]


class AttachmentsBlock(BaseModel):
    count: int
    preview: list[AttachmentRead]


class ReportsBlock(BaseModel):
    count: int
    preview: list[ReportResponse]


class DeadlinesBlock(BaseModel):
    # All deadlines including not_applicable — matches the header KPI's "n/N met"
    # (the donut's deadline wedge, which excludes not_applicable, lives on
    # `completeness.deadlines`). `preview` is the full small list (seeds the
    # deadlines tab / section without a second fetch).
    total: int
    met: int
    overdue: int
    preview: list[DeadlineRead]


class OverviewStats(BaseModel):
    """Header KPIs. `holdback_pct` is the dossier-only required percentage (what
    the header's HOLDBACK chip shows), not the overall donut percentage."""

    deadlines_met: int
    deadlines_total: int
    attachments_count: int
    holdback_pct: int
    delivery_days_remaining: int | None


class ProjectOverviewRead(BaseModel):
    project: ProjectRead
    completeness: CompletenessBlock
    stats: OverviewStats
    findings: FindingsBlock
    certificates: CertificatesBlock
    attachments: AttachmentsBlock
    reports: ReportsBlock
    deadlines: DeadlinesBlock
    members: list[ProjectMemberRead]
    activity_timeline: list[ActivityTimelineBucket]
